src/nvim/terminal.c

```diff
@@ -724,7 +724,11 @@ void terminal_paste(long count, char **y_array, size_t y_size)
     for (size_t j = 0; j < y_size; j++) {
       if (j) {
         // terminate the previous line
+#ifdef MSWIN
+        terminal_send(curbuf->terminal, "\r\n", 2);
```

Comment:
Do the tests correctly fail if this is changed to `"\n", 1` ?

Reply:
No it doesn't.  I don't know why.  Any idea how to fix the test?

(I don't have busted installed, so I had to test with [CI][1].)

[1]: https://github.com/3N4N/neovim/actions/runs/4376063722/jobs/7657639577#step:9:230

Reply:
These tests use https://github.com/neovim/neovim/blob/64d3f68c07f7e67bbbd5db11b289705f7569b949/test/functional/terminal/helpers.lua#L54

Maybe `feed_command()` in the tests isn't sending multi-line input, or maybe `tty-test.c` is not being "strict" enough about CRLF vs LF: https://github.com/neovim/neovim/blob/64d3f68c07f7e67bbbd5db11b289705f7569b949/test/functional/fixtures/tty-test.c#L128-L132

Actually, it looks like `tty-test.c` is being invoked once per line, which does not test what we want. Confirmed by this change:

```diff
diff --git a/test/functional/fixtures/tty-test.c b/test/functional/fixtures/tty-test.c
index 6ee77150213d..81398cb77768 100644
--- a/test/functional/fixtures/tty-test.c
+++ b/test/functional/fixtures/tty-test.c
@@ -110,6 +110,7 @@ static void read_cb(uv_stream_t *stream, ssize_t cnt, const uv_buf_t *buf)
   };
   uv_write(&req, STRUCT_CAST(uv_stream_t, &out), &b, 1, NULL);
   uv_run(&write_loop, UV_RUN_DEFAULT);
+  fprintf(stderr, "xxxxxxxxx\n");
 
   uv_close(STRUCT_CAST(uv_handle_t, &out), NULL);
   uv_run(&write_loop, UV_RUN_DEFAULT);
```

Reply:
It is clear from the following code that even after calling `set_shell_powershell()`, the command that `thelper.screen_setup()` executes is `tty-test.exe`.

https://github.com/neovim/neovim/blob/8065fc9aaeff734f38109aec52bf852379a5a183/test/functional/terminal/helpers.lua#L54-L58

Therefore, if you want to test pasting into PowerShell, you need to pass `powershell.exe` as the command argument to pass to `thelper.screen_setup()`.

Reply:
Good suggestion, though I think we could also test it by modifying `tty-test` to mimic cmd.exe / powershell behavior?

Reply:
I remember a long time ago that `libuv` used to convert newline codes. Therefore, it is easier to use `PowerShell`.

https://github.com/neovim/neovim/blob/d15abd1be4ae85b10174e3ee139d3b7605e87577/src/nvim/tui/tui.c#L1775

Reply:
> Therefore, if you want to test pasting into PowerShell, you need to pass powershell.exe as the command argument to pass to `thelper.screen_setup()`

I am having trouble implementing this.  First I didn't know how to change an instruction in `before_each()` block because changing it would affect all tests.  I found that I can call `clear()` and reassign a new screen with `screen_setup()`.  Using pwsh was successful.  But pwsh starts with a unique prompt, containing current working directory and other stuff.  I thought I could change the prompt to a simple `PS> `, but my input aren't being accepted.  It's being registered in the screen, I can confirm that with `screenshot_util`, but the pwsh prompt isn't changing.  In fact, the next pwsh prompt doesn't appear at all, no matter how many times I `feed('<cr>')` into it.

I'll try again next week but I don't think I can make any progress.  I don't really know lua neither the test-suite and I can't really spend time on it.  Someone needs to take this PR over from me.

Reply:
I don't think `test_terminal_put_operator()`, etc. can be used as is for `PowerShell`. Therefore, why not try using a separate 
`describe` clause such as the following  (I haven't tried it to see if it works because the MinGW test is broken and I haven't actually run it)? I'm doing something weird like `[char]0x3E` due to quoting, but the prompt should be `>`.

```lua
describe('powershell on terminal', function()
  local screen

  before_each(function()
    clear()
    feed_command('set modifiable swapfile undolevels=20')
    poke_eventloop()
    local command = '["' .. is_os('win') and 'powershell' or 'pwsh' .. '", "-NoLogo", "-NoProfile", "-NoExit", "-Command", "function Prompt{[char]0x3E}"]'
    screen = thelpers.screen_setup(nil, command)
  end)

  it('":put" command sends data normally', function()
    feed('<c-\\><c-n>gg"ayj')
    feed_command('let @a = "1\n2\n3\n" . @a')
    feed_command('put a')
    screen:expect([[
    …
    ]])
  end)

  it('send data normally when "put" operator is used', function()
  …
  end)
end)
```

Reply:
Thanks.  I didn't realize I can pass args in the list/dictionary of `command`.  Massively useful.

Unfortunately, the results of `:put a` or `"ap` aren't what's expected.

Here's the snapshot of `:put a`.  "1\n2\n3" doesn't appear on screen.
```
screen:expect{grid=[[
                                                    |
  {2: }                                                 |
                                                    |
                                                    |
                                                    |
  ^                                                  |
  :put a                                            |
]]}
```

If I remove `feed('<c-\\><c-n>gg"ayj')` from your example, the output contains more stuff but not "1\n2\n3":

```
screen:expect{grid=[[
  :let @a = "1                                      |
  2                                                 |
  3                                                 |
  "                                                 |
  :put a                                            |
  {1: }                                                 |
  {3:-- TERMINAL --}                                    |
]]}
```

You might think the @a has been put somewhere above visible screen.  But I extended `extra_rows` arg of `screen_setup()` and it just shows empty rows below, not "1\n2\n3".

Any idea?

Reply:
Wait.  This may be a different problem.  `feed_command('let @a = "1\n2\n3\n"')` doesn't work properly.  It gets `E114: Missing quote: "1`.  I think maybe because it's literally putting a newline in the command.

Reply:
Any idea how to `screen:expect` `{UNEXPECTED foreground = tonumber('0x000002')`?  This is what powershell is emitting.

Reply:
It seems to be possible to pass different `cmd` to `before_each` for each `it` in the following way without having to separate the `describe` clause. This may not be useful information as it seems to have been implemented already, but it is for your information.

```lua
describe('Test', function()
  local cmd = nil
  before_each(function() print('before ' .. tostring(cmd)) end)

  it('1', function()
    assert.is_equal(1, 1)
    cmd = 'powershell'
  end)

  it('2', function()
    assert.is_equal(1, 1)
    cmd = 'pwsh'
  end)

  it('3', function()
    assert.is_equal(1, 1)
  end)
end)
```

result

```
before nil
●before powershell
●before pwsh
●
3 successes / 0 failures / 0 errors / 0 pending : 0.001516 seconds
```

Reply:
Thank you.  It's useful.  (Although, I am confused as to how it works this way, but doesn't matter.)